connectors/bitbucket-server/src/main/java/com/vmware/connectors/bitbucket/server/BitbucketServerController.java

```diff
@@ -0,0 +1,483 @@
+                                    )
+                            )
+                            .toUri();
+
+                    // This header is added for skipping the CSRF check. Otherwise, Atlassian denies the pull request ACTION.
+                    headers.add(ATLASSIAN_TOKEN, "no-check");
+
+                    final ListenableFuture<ResponseEntity<String>> response = this.rest.exchange(
+                            uri,
+                            HttpMethod.POST,
+                            new HttpEntity<>(headers),
+                            String.class);
+
+                    return Async.toSingle(response);
+                });
+
+    }
+
+    private HttpHeaders buildHeaders(final String authHeader) {
+        final HttpHeaders headers = new HttpHeaders();
+        headers.add(AUTHORIZATION, authHeader);
+        return headers;
+    }
+
+    private Single<String> getVersion(final HttpHeaders headers,
+                              final String baseUrl,
+                              final BitbucketServerPullRequest pullRequest) {
+
+        final ListenableFuture<ResponseEntity<JsonDocument>> response = getPullRequestInfo(headers, pullRequest, baseUrl);
+        return Async.toSingle(response)
+                .map(entity -> entity.getBody())
+                .map(jsonDocument -> Integer.toString(jsonDocument.read("$.version")));
+
+    }
+
+    private Observable<Card> getCardsForBitbucketServerPR(final HttpHeaders headers,
+                                                          final BitbucketServerPullRequest pullRequest,
+                                                          final String baseUrl,
+                                                          final String routingPrefix) {
+        logger.debug("Requesting pull request info from bitbucket server base url: {} and pull request info: {}", baseUrl, pullRequest);
+
+        final Single<ResponseEntity<JsonDocument>> bitBucketServerResponse = Async.toSingle(getPullRequestInfo(headers, pullRequest, baseUrl));
+        final Single<List<String>> comments = getComments(baseUrl, headers, pullRequest);
+
+        return Observable.zip(bitBucketServerResponse.toObservable(), comments.toObservable(), Pair::of)
+                .onErrorResumeNext(ObservableUtil::skip404)
+                .map(pair -> convertResponseIntoCard(pair.getLeft(), pullRequest, routingPrefix, pair.getRight()));
+    }
+
+    private ListenableFuture<ResponseEntity<JsonDocument>> getPullRequestInfo(final HttpHeaders headers,
+                                                                              final BitbucketServerPullRequest pullRequest,
+                                                                              final String baseUrl) {
+        final URI uri = UriComponentsBuilder.fromHttpUrl(baseUrl + "/rest/api/1.0/projects/{PROJECT_KEY}/repos/{REPOSITORY_SLUG}/pull-requests/{PULL_REQUEST_ID}")
```

Comment:
Why not just use the `objects` parameter of the `exchange` method instead of building the URI like this? Ditto for the other examples.

Reply:
This can be cleaned up in a few other connectors as well (not in this PR).  We should create a ticket to fix the others.

Reply:
I created APF-783 for this.

Reply:
The one thing that sucks about the `Object...` arg on `rest.exchange` is that they are positional params instead of named ones like these.  With only a handful or params, it's probably not too bad to use the positional params.

Reply:
It looks like some of the overloads also have a Map for the uriVariables.  I remember having trouble getting that one to match what I wanted.  Maybe the overloads just got confusing and I gave up.

I'm not a fan of overloads btw :/

Reply:
I am using the url directly in rest template and passing the date as objects paramater.